drivers/SmartThings/philips-hue/src/handlers/attribute_emitters.lua

```diff
@@ -125,30 +112,175 @@ local function _emit_light_events_inner(light_device, light_repr)
   end
 end
 
+function AttributeEmitters.connectivity_update(child_device, zigbee_status)
+  if zigbee_status.status == "connected" then
+    child_device.log.info_with({hub_logs=true}, "Light status event, marking device online")
+    child_device:online()
+    child_device:set_field(Fields.IS_ONLINE, true)
+  elseif zigbee_status.status == "connectivity_issue" then
+    child_device.log.info_with({hub_logs=true}, "Light status event, marking device offline")
+    child_device:set_field(Fields.IS_ONLINE, false)
+    child_device:offline()
+  end
+end
+
+function AttributeEmitters.emit_button_attribute_events(button_device, button_info)
+  if button_device == nil or (button_device and button_device.id == nil) then
+    log.warn("Tried to emit attribute events for a device that has been deleted")
+    return
+  end
+
+  if button_info.power_state then
+    log.debug(true, "emit power")
+    button_device:emit_event(
+      capabilities.battery.battery(
+        st_utils.clamp_value(button_info.power_state.battery_level, 0, 100)
+      )
+    )
+  end
+
+  local button_idx_map = button_device:get_field(Fields.BUTTON_INDEX_MAP)
+  if not button_idx_map then
+    log.error(
+      string.format(
+        "Button ID to Button Index map lost, " ..
+        "cannot find componenet to emit attribute event on for button [%s]",
+        (button_device and button_device.lable) or "unknown button"
+      )
+    )
+  end
+
+  local idx = button_idx_map[button_info.id] or 1
+  local component_idx
+  if idx == 1 then
+    component_idx = "main"
+  else
+    component_idx = string.format("button%s", idx)
+  end
+
+  local button_report = button_info.button.button_report or { event = "" }
+
+  if button_report.event == "long_press" and not button_device:get_field("button_held") then
+    button_device:set_field("button_held", true)
+    button_device.profile.components[component_idx]:emit_event(
+      capabilities.button.button.held({state_change = true})
+    )
+  end
+
+  if button_report.event == "long_release" and button_device:get_field("button_held") then
+    button_device:set_field("button_held", false)
+  end
+
+  if button_report.event == "short_release" and not button_device:get_field("button_held") then
```

Comment:
So just to check, in a momentary press, the `long_press` would never come in.  Is there a `short_press` equivalent?  I still think this is the right thing here to just generate on the release, but was curious,

Reply:
Yeah, there's no short_press equivalent.

The five supported values are:

- `initial_press`
- `short_release`
- `long_press`
- `repeat`
- `long_release`

The potential event sequences are either:

1. `initial_press` -> `short_release`, or
2. `initial_press` -> `long_press` -> `repeat` -> … -> `long_release`

Reply:
Is `repeat` just sent every second or something if the button is held for a long time?

Reply:
The button service has a `repeat_interval` property that lets you know how often it'll be sent, but yes:

```json
{
  "errors": [],
  "data": [
    {
      "id": "8b5e4b4e-ee4b-4102-9788-1d3b29a859cc",
      "id_v1": "/sensors/9",
      "owner": {
        "rid": "621164bd-30e5-4e4c-9207-024e8f04b643",
        "rtype": "device"
      },
      "metadata": {
        "control_id": 1
      },
      "button": {
        "last_event": "short_release",
        "button_report": {
          "updated": "2024-05-02T20:22:46.673Z",
          "event": "short_release"
        },
        "repeat_interval": 800,
        "event_values": [
          "initial_press",
          "repeat",
          "short_release",
          "long_release",
          "long_press"
        ]
      },
      "type": "button"
    }
  ]
}
```